### PR TITLE
`add_{hline,vline,hrect,vrect}` and `add_{shape,image,annotation}` to multiple facets, squashed

### LINKED ISSUE #2140: add_shape, add_layout_image, add_annotation for multiple facets at once
Right now if I want to add the same shape (i.e. a horizontal line) to every facet in a PX figure, I have to loop through and do it myself. It seems like having a way of saying `row="*"` or something might be handy.

### LINKED ISSUE #2141: add_hline and add_vline (+add_vrect/add_hrect)
We should add these convenience methods, as the ref=x/ref=paper trick is pretty hard for people to understand.

~We'd want to be able to add these to all subplots, one subplot, or possibly a single row or a single column.~ Moved to #2140

We'd also want add_vspan and add_hspan (or vrect/hrect, not set on the names)

### PR DESCRIPTION
**Pull request by nicholas-esterer**
Closes #2141 and closes #2140

### CODE REVIEW

File: doc/python/shapes.md [hunk 1]
```diff
@@ -803,4 +862,32 @@ fig.show(config={'modeBarButtonsToAdd':['drawline',
 ```
 
 ### Reference
-See https://plotly.com/python/reference/layout/shapes/ for more information and chart attribute options!
\ No newline at end of file
+See https://plotly.com/python/reference/layout/shapes/ for more information and chart attribute options!
+
+<!-- #region -->
```
Comment (nicolaskruchten) [inline doc/python/shapes.md:867]: this "what about Dash" section gets added automatically at build time, and so should not be included here please

File: doc/python/horizontal-vertical-shapes.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    thumbnail: thumbnail/horizontal-vertical-shapes.jpg
+---
+
+### Horizontal and Vertical Lines and Boxes (Autoshapes) in Plotly.py
+
+Horizontal and vertical lines and rectangles (autoshapes) that span an entire
+plot can be added via the `add_hline`, `add_vline`, `add_hrect`, and `add_vrect`
+methods of `plotly.graph_objects.Figure`. These shapes are fixed to the
+endpoints of one axis, regardless of the range of the plot, and fixed to data
+coordinates on the other axis. For example
```
Comment (emmanuelle) [inline doc/python/horizontal-vertical-shapes.md:41]:
```suggestion
coordinates on the other axis. Shapes added with these methods are added as [layout shapes](/python/shapes) (as shown when doing `print(fig)`, for example). For example
```

File: doc/python/horizontal-vertical-shapes.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    language: python
+    layout: base
+    name: Shapes
+    order: 37
+    permalink: python/horizontal-vertical-shapes/
+    thumbnail: thumbnail/horizontal-vertical-shapes.jpg
+---
+
+### Horizontal and Vertical Lines and Boxes (Autoshapes) in Plotly.py
+
```
Comment (emmanuelle) [inline doc/python/horizontal-vertical-shapes.md:36]:
```suggestion

*introduced in plotly 4.12*
```

**Comment by nicolaskruchten** [comment]
Needs some changelog entries

**Comment by nicolaskruchten** [comment]
I'm not seeing any shapes rendered when I do this:

```python
import plotly.express as px

df = px.data.tips()
fig = px.scatter(df, x="total_bill", y="tip", color="sex", facet_row="time", facet_col="day")
fig.add_shape(x0=1, x1=1, y0=1, y1=1, row="all", col="all")
fig.show()
```

normal?

**Comment by nicolaskruchten** [comment]
Argh, please ignore last message :)

**Comment by nicolaskruchten** [comment]
The behaviour of `add_trace()` with either or both of `row`/`col` = `"all"` will need to:

1. implement the "only on subplots with pre-existing traces" logic from `.add_hline()` and friends
2. implement the logic we discussed out loud whereby by default all the traces are in the same legendgroup (some UID) unless `legendgroup` is explicitly set to `None` and all but the first trace have `showlegend=False`, unless `showlegend` is explicitly provided.

**Comment by nicolaskruchten** [comment]
Ah, re point 2 above, we [did discuss this](https://github.com/plotly/plotly.py/issues/2140#issuecomment-698975114) and I think I had landed on the idea of documenting the following approach, which is a little verbose, but already works:

```python
fig.add_trace(go.Scatter(x=[1,10], y=[1,5], legendgroup="hi", showlegend=False), row="all", col="all")
fig.data[-1].update(showlegend=True)
```

**Comment by nicolaskruchten** [comment]
Re point 1 above and the same issue for `add_shape` ... [we had discussed a kwarg `exclude_empty_subplots`](https://github.com/plotly/plotly.py/pull/2558#issuecomment-646696273) but this doesn't seem to have survived in this PR :) Man, June seems far away now.

File: doc/python/horizontal-vertical-shapes.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+```
+The default `row` and `col` values are `"all"` so
+`fig.add_vline(x=30, line_color="purple")` is equivalent
+to `fig.add_vline(x=30, row="all", col="all", line_color="purple")` in the above
+example.
+
+#### Adding Text Annotations to Autoshapes
+
+Text can be added to an autoshape using the `annotation` keyword. Using the
+above example:
```
Comment (emmanuelle) [inline doc/python/horizontal-vertical-shapes.md:92]: could you please add a link to the annotations tutorial?

**Review by emmanuelle — COMMENTED** [review]
I'm leaving small comments as I'm reading the code... I have not finished yet. The new domain positioning feature is really cool :-).

File: packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+# _indexing_combinations tests most cases of the following function
+# we just need to test that setting rows or cols to 'all' makes product True,
+# and if not, we can still set product to True.
+@pytest.mark.parametrize(
+    "test_input,expected",
+    [
+        (
+            ("all", [2, 4, 5], False),
+            _unzip_pairs([(r, c) for r in range(1, NROWS + 1) for c in [2, 4, 5]]),
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py:159]:
```suggestion
            zip(*itertools.product(range(1, NROWS + 1), [2, 4, 5]))
```

Comment (emmanuelle) [reply packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py:159]: I think the above does the same as your custom `_unzip_pairs` function (you'd have to import `itertools` as well, it's part of the standard library), so if it's possible to reduce the amount of home-written code it's better for maintenance and readability.

File: doc/python/horizontal-vertical-shapes.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+```python
+import plotly.express as px
+import plotly.graph_objects as go
+
+df = px.data.tips()
+fig = px.scatter(df, x="total_bill", y="tip", facet_row="smoker", facet_col="sex")
+# Add annotations anchored to the top right corner of the resulting lines
+fig.add_vline(x=30, line_color="purple", annotation=go.layout.Annotation(text="A"))
```
[the remark this answers is not in the document]
Comment (nicolaskruchten) [reply doc/python/horizontal-vertical-shapes.md:100]: yeah let's use a dict here please.

**Comment by nicolaskruchten** [comment]
So just to summarize what I think is missing here: the `exclude_empty_subplots` kwarg should exist in add_{shape|annotation|layout_image|trace|hline|vline|hrect|vrect}` and should default to `False` for the methods that existed before and to `True` for the new methods.

File: doc/python/horizontal-vertical-shapes.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicholas-esterer) [reply doc/python/horizontal-vertical-shapes.md:100]: Wait, this was one of the specifications, that it should accept a `go.layout.Annotation` object, as well as a dict, this is to show that fact.

File: packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py (hunk 5 above)
Comment (nicholas-esterer) [reply packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py:159]: Nice, I like this trick!

File: doc/python/horizontal-vertical-shapes.md (hunk 4 above)
Comment (nicholas-esterer) [reply doc/python/horizontal-vertical-shapes.md:92]: Good idea, sure! Is this how you do it? `[annotations](/python/text-and-annotations)`

File: doc/python/images.md [hunk 7]
```diff
@@ -359,5 +359,8 @@ fig.show(config={'modeBarButtonsToAdd':['drawline',
                                        ]})
 ```
 
+
+### Images Placed Relative to the Axes
+
```
[the remark this answers is not in the document]
Comment (nicholas-esterer) [reply doc/python/images.md:364]: Yes, woops ;)

File: packages/python/plotly/plotly/basedatatypes.py [hunk 8]
```diff
@@ -3411,6 +3678,206 @@ def _index_is(iterable, val):
...
+        return self
+
+    add_hrect.__doc__ = _axis_spanning_shapes_docstr("hrect")
+
+    def _has_subplots(self):
+        """ Returns True if figure contains subplots, otherwise it contains a
+        single plot and so this returns False. """
+        return self._grid_ref is not None
+
+    def _subplot_contains_trace(self, xref, yref):
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/basedatatypes.py:3867]: Right now a subplot with no traces but with a shape *does* render and "isn't empty" so I'm sort of thinking this should extend to `shapes` and `annotations` and `images` as well...

**Comment by nicolaskruchten** [comment]
gotta fix the tests :)

File: packages/python/plotly/plotly/basedatatypes.py (hunk 8 above)
Comment (nicholas-esterer) [reply packages/python/plotly/plotly/basedatatypes.py:3867]: you mean we should use a function `_subplot_contains_shape` or something like that? Only add to plots with shapes already on them?

**Comment by nicolaskruchten** [comment]
💃 let's merge this thing!

**Pull request merged**